presto-main/src/main/java/com/facebook/presto/ttl/fetcherManager/ConfidenceBasedTTLFetcherManager.java

```diff
@@ -0,0 +1,191 @@
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        nodeManager.removeNodeChangeListener(nodeChangeListener);
+        if (refreshTTLExecutor != null) {
+            refreshTTLExecutor.shutdownNow();
+        }
+    }
+
+    @VisibleForTesting
+    public synchronized void refreshTTLInfo()
+    {
+        AllNodes allNodes = nodeManager.getAllNodes();
+        refreshTTLInfo(allNodes);
+    }
+
+    private synchronized void refreshTTLInfo(AllNodes allNodes)
+    {
+        Set<InternalNode> activeWorkers = Sets.difference(allNodes.getActiveNodes(), allNodes.getActiveResourceManagers());
+
+        if (!isWorkScheduledOnCoordinator) {
+            activeWorkers = Sets.difference(activeWorkers, allNodes.getActiveCoordinators());
+        }
+
+        Map<NodeInfo, InternalNode> internalNodeMap = activeWorkers
+                .stream()
+                .collect(toImmutableMap(node -> new NodeInfo(node.getNodeIdentifier(), node.getHost()), Function.identity()));
+
+        Map<NodeInfo, NodeTTL> ttlInfo = ttlFetcher.get().getTTLInfo(ImmutableSet.copyOf(internalNodeMap.keySet()));
+
+        nodeTTLMap.putAll(ttlInfo.entrySet().stream().collect(toImmutableMap(e -> internalNodeMap.get(e.getKey()), Map.Entry::getValue)));
+
+        Set<InternalNode> deadNodes = difference(nodeTTLMap.keySet(), activeWorkers).immutableCopy();
+        nodeTTLMap.keySet().removeAll(deadNodes);
+
+        log.info("TTLs refreshed, nodeTtlMap: %s", nodeTTLMap);
+    }
+
+    public Optional<NodeTTL> getTTLInfo(InternalNode node)
+    {
+        return nodeTTLMap.containsKey(node) ? Optional.of(nodeTTLMap.get(node)) : Optional.empty();
+    }
+
+    @Override
+    public Map<InternalNode, NodeTTL> getAllTTLs()
+    {
+        return ImmutableMap.copyOf(nodeTTLMap);
+    }
+
+    @Override
+    public void addTTLFetcherFactory(TTLFetcherFactory ttlFetcherFactory)
+    {
+        requireNonNull(ttlFetcherFactory, "ttlFetcherFactory is null");
+
+        if (ttlFetcherFactories.putIfAbsent(ttlFetcherFactory.getName(), ttlFetcherFactory) != null) {
+            throw new IllegalArgumentException(format("TTL Fetcher factory '%s' is already registered", ttlFetcherFactory.getName()));
+        }
+    }
+
+    @Override
+    public void loadTTLFetcher()
+            throws Exception
+    {
+        String factoryName = "infinite";
+        Map<String, String> properties = ImmutableMap.of();
+
+        if (TTL_FETCHER_CONFIG.exists()) {
+            properties = new HashMap<>(loadProperties(TTL_FETCHER_CONFIG));
+            factoryName = properties.remove(TTL_FETCHER_PROPERTY_NAME);
+
+            checkArgument(!isNullOrEmpty(factoryName),
+                    "TTL Fetcher configuration %s does not contain %s", TTL_FETCHER_CONFIG.getAbsoluteFile(), TTL_FETCHER_PROPERTY_NAME);
+        }
+
+        load(factoryName, properties);
+
+        if (ttlFetcher.get().needsPeriodicRefresh()) {
```

Comment:
Can we simply annotate an internal method on the fetcher with `@Init`, which will ensure it's executed on startup?

Reply:
Sorry, I didn't quite understand the suggestion here. Are you saying we should add an `@Init` annotation to a method in a `TTLFetcher` implementation?